subprojects/core/src/main/java/org/gradle/api/internal/collections/IterationOrderRetainingSetElementSource.java

```diff
@@ -16,75 +16,140 @@
 
 package org.gradle.api.internal.collections;
 
+import com.google.common.base.Objects;
 import org.gradle.api.Action;
+import org.gradle.api.internal.provider.AbstractProvider;
 import org.gradle.api.internal.provider.ProviderInternal;
 
+import javax.annotation.Nullable;
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.HashSet;
 import java.util.Iterator;
 import java.util.LinkedHashSet;
+import java.util.List;
 import java.util.Set;
 
 public class IterationOrderRetainingSetElementSource<T> implements ElementSource<T> {
-    private final Set<T> values = new LinkedHashSet<T>();
+    // This set represents the elements which have been realized and "added" to the store.  This may
+    // or may not be in the order that elements are inserted.  We track this to know which realized
+    // elements have already had rules fired against them.
+    private final Set<T> added = new HashSet<T>();
+
+    // This set represents the order in which elements are inserted to the store, either actual
+    // or provided.  We construct a correct iteration order from this set.
+    private final List<RealizableProvider<T>> inserted = new ArrayList<RealizableProvider<T>>();
+
+    // Represents the pending elements that have not yet been realized.
     private final PendingSource<T> pending = new DefaultPendingSource<T>();
 
+    // Cache the ordered set of values so that we don't have to construct the list every time we need to iterate
+    private final Set<T> orderedValuesCache = new LinkedHashSet<T>();
```

Comment:
It feels like we're duplicating a bunch of state. Each of the providers in `inserted` knows whether it has been realized or not and if so, what its value is, and so the state in `added`, `pending` and orderedValuesCache` is derived from this.

I suspect `orderedValuesCache` at least could be turned into a wrapper object that delegates to `added` and `pending`, so that we don't need to retain another linked hash set in memory. For example, for iteration it can just iterate over `pending` and skip those things that aren't realized.

There's probably some other things we could do to reduce the duplication.

Reply:
Absolutely.  I originally wrote this with just the `inserted` collection, but introduced `orderedValuesCache` for performance reasons.  Basically, when we add something, trying to determine if what we are adding is a new value or an existing just-realized value is expensive if we have to iterate over `inserted` to answer that question.  I'll go back to that and see if I can find a performance optimization that doesn't require an extra hash set.

Reply:
Ok, I changed this back to using the `inserted` collection as the source for everything and I think it looks like I've been able to mitigate some of the things I was doing before that were causing a performance regression.  The implementation is much simpler now, too, which is good.